src/components/Modal/Modal.tsx

```diff
@@ -0,0 +1,160 @@
+import React, { CSSProperties, useCallback, useEffect, useMemo } from 'react';
+import { CSSTransition } from 'react-transition-group';
+import classNames from 'classnames';
+import { ClassValue } from 'classnames/types';
+
+import ReactDOM from 'react-dom';
+import styles from './Modal.module.css';
+
+type Size = 'sm' | 'md' | 'lg';
+type Level = 'primary' | 'warning' | 'info' | 'danger';
+
+interface ModalProps {
+  children?: React.ReactNode;
+  className?: ClassValue;
+  isOpen?: boolean;
+  onClose?: () => void;
+  size?: Size;
+  style?: CSSProperties;
+}
+
+type ModalChildProps = Pick<ModalProps, 'children' | 'className'>;
+type ModalHeaderProps = Pick<ModalProps, 'children' | 'className'> & {
+  level?: Level;
+};
+
+type ModalType = React.FC<ModalProps> & {
+  Header: React.FC<ModalHeaderProps>;
+  Footer: React.FC<ModalChildProps>;
+  Body: React.FC<ModalChildProps>;
+};
+
+const Modal: ModalType = ({
+  children,
+  className: passedClassName = '',
+  isOpen = false,
+  onClose = () => {},
+  size = 'md',
+  style: passedStyle = {},
+  ...passedProps
+}) => {
+  const handleEscape = useCallback(
+    (e: KeyboardEvent) => {
+      if (isOpen && e.code === 'Escape' && onClose) {
+        onClose();
+      }
+    },
+    [isOpen, onClose]
+  );
+
+  useEffect(() => {
+    if (isOpen) {
+      window.addEventListener('keyup', handleEscape);
+    }
+
+    if (!isOpen) {
+      window.removeEventListener('keyup', handleEscape);
+    }
+
+    return () => {
+      if (isOpen) window.removeEventListener('keyup', handleEscape);
+    };
+  }, [handleEscape, isOpen]);
+
+  useEffect(() => {
```

Comment:
This is probably good for a follow-up PR:

It would be nice to autofocus inside the modal when it opens, and add a focus trap.

Here's the general idea of the focus trap: https://medium.com/@im_rahul/focus-trapping-looping-b3ee658e5177
And here's an implementation of getting all focusable elements inside a div or the document: https://zellwk.com/blog/keyboard-focusable-elements/

There's a catch mentioned about focus trapping with screen readers, and I don't have a solution for that, but just implementing the auto-focus would still be really useful.